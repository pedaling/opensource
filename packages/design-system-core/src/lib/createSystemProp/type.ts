import type { Theme } from '../types';

export type InterpolationFn = (props: Record<string, any>) => Record<string, any>;

export type SystemPropConfig = {
  property: string;
  styleProperty?: string;
  scale?: keyof Pick<Theme, 'colors' | 'space' | 'opacity'>;
  disabled?: boolean;
  transform?: (value: any) => Record<string, any> | undefined;
};

export type SystemProp = {
  (props: Record<string, any>): Record<string, any>;
  propName: string;
  disabled: boolean;
};
