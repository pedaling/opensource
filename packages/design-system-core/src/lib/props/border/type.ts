import type { ResponsiveValue } from '../../types';

export type BorderProps = {
  borderWidth?: ResponsiveValue<number>;
  borderStyle?: ResponsiveValue<'solid' | 'none'>;
  borderColor?: ResponsiveValue<string>;
  borderTopWidth?: ResponsiveValue<number>;
  borderTopStyle?: ResponsiveValue<'solid' | 'none'>;
  borderTopColor?: ResponsiveValue<string>;
  borderRightWidth?: ResponsiveValue<number>;
  borderRightStyle?: ResponsiveValue<'solid' | 'none'>;
  borderRightColor?: ResponsiveValue<string>;
  borderBottomWidth?: ResponsiveValue<number>;
  borderBottomStyle?: ResponsiveValue<'solid' | 'none'>;
  borderBottomColor?: ResponsiveValue<string>;
  borderLeftWidth?: ResponsiveValue<number>;
  borderLeftStyle?: ResponsiveValue<'solid' | 'none'>;
  borderLeftColor?: ResponsiveValue<string>;
  borderRadius?: ResponsiveValue<number | string>;
  borderTopLeftRadius?: ResponsiveValue<number | string>;
  borderTopRightRadius?: ResponsiveValue<number | string>;
  borderBottomLeftRadius?: ResponsiveValue<number | string>;
  borderBottomRightRadius?: ResponsiveValue<number | string>;
  outlineWidth?: ResponsiveValue<number>;
  outlineStyle?: ResponsiveValue<'solid' | 'none'>;
  outlineColor?: ResponsiveValue<string>;
  outlineOffset?: ResponsiveValue<number>;
};
