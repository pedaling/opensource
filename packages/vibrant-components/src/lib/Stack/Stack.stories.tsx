import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { Box } from '@vibrant-ui/core';
import { Stack } from './Stack';

export default {
  title: 'Stack',
  component: Stack,
  args: {
    direction: 'vertical',
    spacing: 20,
  },
} as ComponentMeta<typeof Stack>;

export const Basic: ComponentStory<typeof Stack> = props => (
  <Stack {...props}>
    <Box width={200} height={200} backgroundColor="primary" borderRadius={12} />
    <Box width={200} height={200} backgroundColor="informative" borderRadius={12} />
  </Stack>
);

export const Scrollable: ComponentStory<typeof Stack> = props => (
  <Stack height={300} width={300} scrollable={true} {...props}>
    <Box width={200} height={200} backgroundColor="primary" borderRadius={12} flexShrink={0} />
    <Box width={200} height={200} backgroundColor="informative" borderRadius={12} flexShrink={0} />
  </Stack>
);
