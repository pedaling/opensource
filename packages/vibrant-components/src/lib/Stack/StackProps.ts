import type { RefObject } from 'react';
import type {
  DisplayProps,
  FlexboxProps,
  OverflowProps,
  ReactElementChild,
  ResponsiveValue,
  SizingProps,
  SpacingProps,
} from '@vibrant-ui/core';
import { propVariant, withVariation } from '@vibrant-ui/core';

type SemanticTagName = 'article' | 'aside' | 'footer' | 'header' | 'nav' | 'section';

export type StackProps = DisplayProps &
  FlexboxProps &
  OverflowProps &
  SizingProps &
  SpacingProps & {
    as?: SemanticTagName | 'div' | 'label';
    direction: ResponsiveValue<'horizontal' | 'vertical'>;
    ref?: RefObject<HTMLElement>;
    spacing?: ResponsiveValue<number>;
    children?: ReactElementChild | ReactElementChild[];
  };

export const withStackVariation = withVariation<StackProps>('Stack')(
  propVariant({
    props: [
      {
        name: 'direction',
        responsive: true,
        keep: true,
      },
    ],
    variants: {
      horizontal: {
        flexDirection: 'row',
      },
      vertical: {
        flexDirection: 'column',
      },
    } as const,
  }),
  propVariant({
    props: [
      {
        name: 'direction',
        responsive: true,
      },
      {
        name: 'spacing',
        responsive: true,
        keep: true,
      },
    ],
    variants: ({ direction, spacing }) => {
      if (!spacing) {
        return {
          spaceWidth: 0,
          spaceHeight: 0,
        };
      }

      if (direction === 'horizontal') {
        return {
          spaceWidth: spacing,
          spaceHeight: 0,
        };
      }

      return {
        spaceWidth: 0,
        spaceHeight: spacing,
      };
    },
  })
);
