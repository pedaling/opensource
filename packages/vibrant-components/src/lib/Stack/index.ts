export { Stack } from './Stack';
export type { StackProps, Alignment } from './StackProps';
