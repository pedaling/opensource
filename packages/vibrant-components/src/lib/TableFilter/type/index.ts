export type StringFilterOperator = 'contains' | 'empty' | 'equals' | 'notContains' | 'notEmpty' | 'notEquals';

export type StringFilter = {
  value: string;
  operator: StringFilterOperator[];
};

export type Filter = StringFilter & { dataKey: string };
