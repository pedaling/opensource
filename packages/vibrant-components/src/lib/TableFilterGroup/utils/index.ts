import type { DateFilterOperator, MultiSelectFilterOperator, Option, StringFilterOperator } from '../types';

export const isValueRequiredOperator = (op: string) => op === 'empty' || op === 'notEmpty';

export const isStringFilterValid = (filter: { value: string; operator: StringFilterOperator }) =>
  Boolean(filter.value) || isValueRequiredOperator(filter.operator);

export const isDateFilterValid = (filter: { value: Date[]; operator: DateFilterOperator }) => {
  if (isValueRequiredOperator(filter.operator)) return true;

  if (filter.operator === 'between') return filter.value.length >= 2;

  return filter.value.length >= 1;
};

export const isMultiSelectFilterValid = (filter: { value: Option['value'][]; operator: MultiSelectFilterOperator }) =>
  Boolean(filter.value.length > 0) || isValueRequiredOperator(filter.operator);
