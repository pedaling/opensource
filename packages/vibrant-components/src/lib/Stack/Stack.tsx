import { Children, Fragment, isValidElement } from 'react';
import { Box } from '@vibrant-ui/core';
import { withStackVariation } from './StackProps';

export const Stack = withStackVariation(({ spacing, spaceWidth, spaceHeight, children, ...restProps }) => {
  const validChildren = Children.toArray(children).filter(child => isValidElement(child));

  return (
    <Box display="flex" {...restProps}>
      {spacing
        ? validChildren.map((child, index, { length: childLength }) => (
            <Fragment key={index}>
              {child}
              {childLength !== index + 1 && <Box flexGrow={0} flexShrink={0} width={spaceWidth} height={spaceHeight} />}
            </Fragment>
          ))
        : children}
    </Box>
  );
});
