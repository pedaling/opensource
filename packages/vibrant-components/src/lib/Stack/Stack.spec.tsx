import { Box } from '@vibrant-ui/core';
import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { Stack } from './Stack';

describe('<Stack />', () => {
  const { render } = createReactRenderer();

  let renderer: ReactRenderer;

  let element: HTMLElement;

  describe('when empty stack rendered', () => {
    beforeEach(async () => {
      renderer = render(<Stack data-testid="stack" direction="vertical" />);

      element = await renderer.findByTestId('stack');
    });

    it('div element created', () => {
      expect(element.tagName).toBe('DIV');
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when vertical stack with spacing rendered', () => {
    beforeEach(async () => {
      renderer = render(
        <Stack data-testid="stack" spacing={20} direction="vertical">
          <Box width={200} height={200} />
          <Box width={200} height={200} />
        </Stack>
      );

      element = await renderer.findByTestId('stack');
    });

    it('flex-direction is column', async () => {
      expect(element).toHaveStyleRule('flex-direction', 'column');
    });

    it('spacing element with height exists', async () => {
      expect(element.childElementCount).toBe(3);

      expect(element.children[1]).toHaveStyleRule('width', '0');

      expect(element.children[1]).toHaveStyleRule('height', '20px');
    });

    it('match snapshot', async () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when horizontal stack with spacing rendered', () => {
    beforeEach(async () => {
      renderer = render(
        <Stack data-testid="stack" direction="horizontal" spacing={20}>
          <Box width={200} height={200} />
          <Box width={200} height={200} />
        </Stack>
      );

      element = await renderer.findByTestId('stack');
    });

    it('flex-direction is row', async () => {
      expect(element).toHaveStyleRule('flex-direction', 'row');
    });

    it('spacing element with width exists', async () => {
      expect(element.childElementCount).toBe(3);

      expect(element.children[1]).toHaveStyleRule('width', '20px');

      expect(element.children[1]).toHaveStyleRule('height', '0');
    });

    it('match snapshot', async () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when responsive direction stack rendered', () => {
    beforeEach(async () => {
      renderer = render(
        <Stack data-testid="stack" direction={['vertical', 'horizontal']} spacing={[20, 30]}>
          <Box width={200} height={200} />
          <Box width={200} height={200} />
        </Stack>
      );

      element = await renderer.findByTestId('stack');
    });

    it('vertical stack used in first breakpoint', () => {
      expect(element).toHaveStyleRule('flex-direction', 'column');

      expect(element.children[1]).toHaveStyleRule('width', '0');

      expect(element.children[1]).toHaveStyleRule('height', '20px');
    });

    it('horizontal stack used in second breakpoint', () => {
      expect(element).toHaveStyleRule('flex-direction', 'row', {
        media: '@media screen and (min-width: 640px)',
      });

      expect(element.children[1]).toHaveStyleRule('width', '30px', {
        media: '@media screen and (min-width: 640px)',
      });

      expect(element.children[1]).toHaveStyleRule('height', '0', {
        media: '@media screen and (min-width: 640px)',
      });
    });

    it('match snapshot', async () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when children is falsy value', () => {
    beforeEach(async () => {
      renderer = render(
        <Stack data-testid="stack" direction="vertical" spacing={20}>
          {false}
          {null}
        </Stack>
      );

      element = await renderer.findByTestId('stack');
    });

    it('space not exists', () => {
      expect(element.childElementCount).toBe(0);
    });
  });
});
