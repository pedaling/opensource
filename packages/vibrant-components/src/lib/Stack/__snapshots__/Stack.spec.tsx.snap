// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stack /> when empty stack rendered match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

<div>
  <div
    class="emotion-0"
    data-testid="stack"
  />
</div>
`;

exports[`<Stack /> when horizontal stack with spacing rendered match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  width: 200px;
  height: 200px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
  position: relative;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  width: 20px;
  height: 0;
}

<div>
  <div
    class="emotion-0"
    data-testid="stack"
  >
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-1"
    />
  </div>
</div>
`;

exports[`<Stack /> when responsive direction stack rendered match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

@media screen and (min-width: 640px) {
  .emotion-0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  width: 200px;
  height: 200px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
  position: relative;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  width: 0;
  height: 20px;
}

@media screen and (min-width: 640px) {
  .emotion-2 {
    width: 30px;
    height: 0;
  }
}

<div>
  <div
    class="emotion-0"
    data-testid="stack"
  >
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-1"
    />
  </div>
</div>
`;

exports[`<Stack /> when vertical stack with spacing rendered match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  box-sizing: border-box;
  position: relative;
  width: 200px;
  height: 200px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
  position: relative;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  width: 0;
  height: 20px;
}

<div>
  <div
    class="emotion-0"
    data-testid="stack"
  >
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-1"
    />
  </div>
</div>
`;
